# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h
# Commit: 88a9fe04aaa5
# Full Hash: 88a9fe04aaa582676a84734d67c9069b68a7f3e1
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 5
# Description:
#   Bug 1654112 - Update Audio- and VideoConduit stats API and integration. r=ng,bwc
#   
#   For AudioConduit this patch:
#   - Uses correct threads per upstream, pending changes due to the threading model
#     update that's coming
# ==============================================================================

diff -r 3eae9f9279f6 -r 88a9fe04aaa5 dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h
--- a/dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h	Fri Jan 29 15:38:54 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h	Mon Jan 18 15:28:29 2021 +0100
@@ -229,30 +229,6 @@
 
   virtual void SetSyncGroup(const std::string& group) = 0;
 
-  /**
-   * Functions returning stats needed by w3c stats model.
-   */
-
-  virtual bool GetSendPacketTypeStats(
-      webrtc::RtcpPacketTypeCounter* aPacketCounts) = 0;
-
-  virtual bool GetRecvPacketTypeStats(
-      webrtc::RtcpPacketTypeCounter* aPacketCounts) = 0;
-
-  virtual bool GetRTPReceiverStats(unsigned int* jitterMs,
-                                   unsigned int* cumulativeLost) = 0;
-  virtual bool GetRTCPReceiverReport(uint32_t* jitterMs,
-                                     uint32_t* packetsReceived,
-                                     uint64_t* bytesReceived,
-                                     uint32_t* cumulativeLost,
-                                     Maybe<double>* aOutRttMs) = 0;
-  virtual bool GetRTCPSenderReport(unsigned int* packetsSent,
-                                   uint64_t* bytesSent,
-                                   DOMHighResTimeStamp* aRemoteTimestamp) = 0;
-
-  virtual Maybe<mozilla::dom::RTCBandwidthEstimationInternal>
-  GetBandwidthEstimation() = 0;
-
   virtual void GetRtpSources(nsTArray<dom::RTCRtpSourceEntry>& outSources) = 0;
 
   virtual uint64_t CodecPluginID() = 0;
@@ -263,8 +239,10 @@
 
   virtual void DeleteStreams() = 0;
 
+  virtual Maybe<RefPtr<AudioSessionConduit>> AsAudioSessionConduit() = 0;
   virtual Maybe<RefPtr<VideoSessionConduit>> AsVideoSessionConduit() = 0;
 
+  virtual webrtc::Call::Stats GetCallStats() const = 0;
   virtual void SetRtcpEventObserver(RtcpEventObserver* observer) = 0;
 
   NS_INLINE_DECL_THREADSAFE_REFCOUNTING(MediaSessionConduit)
@@ -391,6 +369,14 @@
 
   Type type() const override { return VIDEO; }
 
+  Maybe<RefPtr<AudioSessionConduit>> AsAudioSessionConduit() override {
+    return Nothing();
+  }
+
+  Maybe<RefPtr<VideoSessionConduit>> AsVideoSessionConduit() override {
+    return Some(RefPtr<VideoSessionConduit>(this));
+  }
+
   /**
    * Function to attach Renderer end-point of the Media-Video conduit.
    * @param aRenderer : Reference to the concrete Video renderer implementation
@@ -456,19 +442,10 @@
 
   bool UsingFEC() const { return mUsingFEC; }
 
-  virtual bool GetVideoEncoderStats(double* framerateMean,
-                                    double* framerateStdDev,
-                                    double* bitrateMean, double* bitrateStdDev,
-                                    uint32_t* droppedFrames,
-                                    uint32_t* framesEncoded,
-                                    Maybe<uint64_t>* qpSum) = 0;
-  virtual bool GetVideoDecoderStats(double* framerateMean,
-                                    double* framerateStdDev,
-                                    double* bitrateMean, double* bitrateStdDev,
-                                    uint32_t* discardedPackets,
-                                    uint32_t* framesDecoded) = 0;
+  virtual Maybe<webrtc::VideoReceiveStream::Stats> GetReceiverStats() const = 0;
+  virtual Maybe<webrtc::VideoSendStream::Stats> GetSenderStats() const = 0;
 
-  virtual void RecordTelemetry() const = 0;
+  virtual void RecordTelemetry() = 0;
 
   virtual bool AddFrameHistory(
       dom::Sequence<dom::RTCVideoFrameHistoryInternal>* outHistories) const = 0;
@@ -503,6 +480,10 @@
 
   Type type() const override { return AUDIO; }
 
+  Maybe<RefPtr<AudioSessionConduit>> AsAudioSessionConduit() override {
+    return Some(this);
+  }
+
   Maybe<RefPtr<VideoSessionConduit>> AsVideoSessionConduit() override {
     return Nothing();
   }
@@ -563,6 +544,9 @@
 
   virtual bool InsertDTMFTone(int channel, int eventCode, bool outOfBand,
                               int lengthMs, int attenuationDb) = 0;
+
+  virtual Maybe<webrtc::AudioReceiveStream::Stats> GetReceiverStats() const = 0;
+  virtual Maybe<webrtc::AudioSendStream::Stats> GetSenderStats() const = 0;
 };
 }  // namespace mozilla
 #endif